shared/src/commonMain/kotlin/io/customer/shared/tracking/api/model/Metric.kt

```diff
@@ -9,7 +9,7 @@ import kotlinx.serialization.SerialName
 @kotlinx.serialization.Serializable
 internal data class Metric(
     @SerialName("delivery_id") val deliveryID: String,
-    @SerialName("device_id") val deviceToken: String,
+    @SerialName("device_id") val deviceToken: String?,
```

Comment:
Reason for this being optional?

Reply:
Because we don't need token for [in-app metrics](https://github.com/customerio/customerio-android/blob/6da8b8d3d16fa1c9c0acdb6012271a2252b30951/sdk/src/main/java/io/customer/sdk/queue/Queue.kt#L46) and we are now using same models for [both metrics](https://github.com/customerio/mobile-shared/blob/945c9cb4ca5e02374ef5f3a305b254dbc337a58a/shared/src/commonMain/kotlin/io/customer/shared/tracking/queue/BackgroundQueue.kt#L150-L179). This, however, is only optional in KMM model but not for the [calls initiated from SDKs](https://github.com/customerio/mobile-shared/blob/945c9cb4ca5e02374ef5f3a305b254dbc337a58a/shared/src/commonMain/kotlin/io/customer/shared/tracking/queue/BackgroundQueue.kt#L52-L63).

Reply:
At a minimum, we should have comments explaining why this value is optional. 

But, I have another idea to propose. 

I think it can be error-prone to have data classes be re-used for multiple purposes but share the same constructor. Instead, what about creating convenient constructors for each use case? 

```kt
Activity.Metric.inApp(...) // this constructor doesn't ask for a deviceToken
Activity.Metric.push(...) // this constructor requires a deviceToken

// But both constructors return the same `Activity.Metric` data type. 
```

Another note, I know that in-app metric *events* are not the same as push metric *events*. 

So, might be worth having a `InAppMetricEvent` and a separate `PushMetricEvent` where each have their appropriate event options in them. We can re-use data classes such as `Metric` and `Activity.Metric`, but if we have separate constructors as talked about above, we could use specialized metric events:

```kt
Activity.Metric.inApp(event: InAppMetricEvent, ...)
Activity.Metric.push(event: PushMetricEvent, ...)
```

Reply:
Makes sense, I'll take a look and update this with one of the suggested solutions.

Reply:
Added named constructors using companion object, but not updating MetricEvent since enums cannot be inherited from enums and this will be a breaking change for customers sending any metric events. Dropping it for now, we can make multiple MetricEvents in future when needed. Added this comment to improvements list.